aesara/scan/basic.py

```diff
@@ -861,7 +861,10 @@ def wrap_into_list(x):
     # when we apply the lambda expression we get a mixture of update rules
     # and outputs that needs to be separated
 
-    condition, outputs, updates = get_updates_and_outputs(fn(*args))
+    with collect_new_shareds() as new_shareds:
+        raw_inner_outputs = fn(*args)
```

Comment:
Wow, this is really clever!!! Do we only need to know if it was called here, or do we need to keep some nested level counter as well?

Asking because the Theano guys were thinking along those lines: https://github.com/Theano/Theano/issues/1565, but perhaps they were overcomplicating

Reply:
> Wow, this is really clever!!! Do we only need to know if it was called here, or do we need to keep some nested level counter as well?

It was the most expedient solution.  In general, I really dislike using these global variable approaches.  

We shouldn't need any sort of nested counter; each time the context manager is used, a new `list` is created to gather the constructed `SharedVariable`s exclusive to that "scope", and, since no "outer-scopes" should be accessing variables in their "inner-scopes", it works out.

> Asking because the Theano guys were thinking along those lines: [Theano/Theano#1565](https://github.com/Theano/Theano/issues/1565), but perhaps they were overcomplicating

Yeah, that looks **wildly** over-complicated for this task, but the underlying notion is correct (and fairly obvious): `SharedVariable`s are inherently "global" in scope, which doesn't work with nested scopes.